File: GeoFirestoreExample/geofirestore/src/main/java/org/imperiumlabs/geofirestore/EventListenerBridge.java
package org.imperiumlabs.geofirestore;

import com.google.firebase.firestore.*;
import java.lang.Exception;

// FULLY TESTED

/**
 * GeoQuery notifies listeners with this interface about documents that entered, exited, or moved within the query.
 */
final class EventListenerBridge implements GeoQueryDataEventListener {
    private final GeoQueryEventListener listener;

    public EventListenerBridge(final GeoQueryEventListener listener) {
        this.listener = listener;
    }

    @Override
    public void onDocumentEntered(final DocumentSnapshot documentSnapshot, final GeoPoint location) {
        listener.onKeyEntered(documentSnapshot.getId(), location);
    }

    @Override
    public void onDocumentExited(final DocumentSnapshot documentSnapshot) {
        listener.onKeyExited(documentSnapshot.getId());
    }

    @Override
    public void onDocumentMoved(final DocumentSnapshot documentSnapshot, final GeoPoint location) {
        listener.onKeyMoved(documentSnapshot.getId(), location);
    }

    @Override
    public void onDocumentChanged(final DocumentSnapshot documentSnapshot, final GeoPoint location) {
        // No-op.
    }

    @Override
    public void onGeoQueryReady() {
        listener.onGeoQueryReady();
    }

    @Override
    public void onGeoQueryError(final Exception exception) {
        listener.onGeoQueryError(exception);
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        final EventListenerBridge that = (EventListenerBridge) o;
        return listener.equals(that.listener);
    }

    @Override
    public int hashCode() {
        return listener.hashCode();
    }
}

File: geofirestore/src/main/java/org/imperiumlabs/geofirestore/listeners/EventListenerBridge.kt
package org.imperiumlabs.geofirestore.listeners

import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.GeoPoint

// FULLY TESTED

/**
 * GeoQuery notifies listeners with this interface about documents that entered, exited, or moved within the query.
 */
class EventListenerBridge(private val listener: GeoQueryEventListener): GeoQueryDataEventListener {

    override fun onDocumentEntered(documentSnapshot: DocumentSnapshot, location: GeoPoint) {
        listener.onKeyEntered(documentSnapshot.id, location)
    }

    override fun onDocumentExited(documentSnapshot: DocumentSnapshot) {
        listener.onKeyExited(documentSnapshot.id)
    }

    override fun onDocumentMoved(documentSnapshot: DocumentSnapshot, location: GeoPoint) {
        listener.onKeyMoved(documentSnapshot.id, location)
    }

    override fun onDocumentChanged(documentSnapshot: DocumentSnapshot, location: GeoPoint) {
        //No-op
    }

    override fun onGeoQueryReady() {
        listener.onGeoQueryReady()
    }

    override fun onGeoQueryError(exception: Exception) {
        listener.onGeoQueryError(exception)
    }


    override fun equals(other: Any?): Boolean {
        if (other == null || other !is EventListenerBridge)
            return false
        return listener == other.listener
    }

    override fun hashCode(): Int {
        return listener.hashCode()
    }
}
